File: Cplusplus/0039_combination_sum/0039_combination_sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> list;
        sort(candidates.begin(), candidates.end());
        dfs(candidates, ans, list, 0, target);
        return ans;
    }

    void dfs(vector<int>& candidates, vector<vector<int>> &ans, vector<int> &list, int index, int target) {
        if (target == 0) {
            ans.push_back(list);
            return;
        }

        for (int i = index; i < candidates.size(); ++i) {
            if (target - candidates[i] < 0) 
                break;
            list.push_back(candidates[i]);
            dfs(candidates, ans, list, i, target - candidates[i]);
            list.pop_back();
        }
    }
};

File: Java/0039_combination_sum/0039_combination_sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        // Arrays.sort(candidates);
        backtrack(res, new ArrayList<Integer>(), candidates, target, 0);
        return res;
    }

    private void backtrack(List<List<Integer>> res, ArrayList<Integer> current, int[] candidates, int target, int index) {
        if (target < 0 || index >= candidates.length)
            return;
        if (target == 0) {
            res.add(new ArrayList<Integer>(current));
            return;
        }
        else {
            current.add(candidates[index]);
            backtrack(res, current, candidates, target - candidates[index], index);
            current.remove(current.indexOf(candidates[index]));
            backtrack(res, current, candidates, target, index + 1);
        }
    }
}